### ISSUE TITLE
Detekt Gradle Task Is Not Remote Gradle Cache Friendly

### ISSUE THREAD

**Issue by AlexanderGH** [detekt/detekt#7170]
## Expected Behavior
It should be possible to configure the detekt tasks in such a way as they can be cached in CI or one machine and then get a cache hit on another machine where the code lives under a different path while not breaking any functionality.

## Observed Behavior
It is not possible to use the gradle remote cache without breaking the basePath functionality since the base path, at least from the documentation, takes an absolute path which will differ on different machines which will then result in different task cache keys and therefore cache misses.

## Steps to Reproduce
Run detekt on two machines where the module has different paths on the file system. If you have the gradle remote cache enabled, you will notice that there is a cache miss and different cache keys for the two runs due ti the different basePath values.

The only way I've found to make it work is to pass "" as the basePath, but then we get non-relativized paths in the generated reports.

## Context
It's nice to allow developers and CI builds to leverage a shared cache.

## Your Environment
* Version of detekt used: 1.23.6
* Version of Gradle used: 8.7
* Gradle scan link (add `--scan` option when running the gradle task): <company policy prohibits sharing>
* Operating System and version: <varies>
* Link to your project (if it's a public repository):

**Comment by BraisGabin** [comment]
I think that to be able to implement this we will need #6879.

With that we could split the task in two. One with the analysis (that is the heavy part) and it should be cacheable and another(s) task(s) with the report creation. That last one would not be cacheable because it would depend on the absolute path but it should be a really fast.

**Comment by AlexanderGH** [comment]
What are your thoughts of accepting a relative path (relative to the module path)?

Then the implementation would do something like:
fileWithIssues.stripPrefix(File(projectDir.absolutePath, baseDir))

I imagine that's the usual use-case and would allow everything to be cacheable.

Alternatively, I think this could be solved by accepting a File instead of a string and marking it with `@PathSensitive(PathSensitivity.NAME_ONLY)`. I would hope gradle would then relativize the input and make the outputs relocatable.

**Comment by AlexanderGH** [comment]
Here's my concrete proposal that should be mostly backwards compatible:
- accept the directory of the project `projectRootPath` as an input with `@PathSensitive(PathSensitivity.NAME_ONLY)` defaulting to the current dir by convention
- if `basePath` does not start with a "/", assume it is relative and then get the base path using `File(projectRootPath, basePath)`
  - You might need to use some level of canonicalPath or absolutePath on the base path and detected file locations
- if `basePath` starts with a '/', assume it is the absolute path and just use that as you do today

**Comment by AlexanderGH** [comment]
Commenting to remove stale tag...

**Comment by 3flex** [comment]
We use basePath for a few things:

* Include/exclude rules. We only care about the relative paths within the project, not the full file path, so we need to define a base path for that to work.
* In some output, e.g. reporting issues on the console. Even though the full file path is displayed, we only need to show it when there are issues to show, and console output isn't cached as an output anyway.
* In MD & HTML reports to show the full path.
* In SARIF report to set [`uriBaseId`](https://docs.oasis-open.org/sarif/sarif/v2.1.0/errata01/os/sarif-v2.1.0-errata01-os-complete.html#_Toc141790681).

basePath can be a relative path for the first two. It can always be relative when used as an input.

What makes sense for the MD & HTML reports? Does the absolute path add value in the output? Maybe we switch to relative paths. Android Lint uses relative by default but allows enabling absolute paths with a flag. That could be a good compromise and allows the basepath input to have relative path sensitivity. We can switch to relative path output and add a flag later if there's demand.

Apparently in SARIF we're wasting our time dealing with absolute paths, GitHub at least uses [its own logic](https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning#specifying-the-root-for-source-files) to deal with absolute paths, so we can strike this off the list. Not sure about other SARIF consumers. But from a producer perspective we're spec compliant if we remove `uriBaseId` from the output and deal only with relative paths. I think we should do this.

@BraisGabin you're close to this, what do you think?

**Comment by BraisGabin** [comment]
Multiple things here:

- If we split the issue finding in one gradle task and the report creation in another task we should be fine. The task that generates the reports will not be cacheable but that shouldn't be a problem, in most cases it should be a final task and it should be really fast. The task that finds the issues is the heavy one and that would work with remote cache.
- On `md` and `html` we don't show the absolute paths. We use the absolute paths to read the files to generate the code snippets.
- About removing `uriBaseId`: If we split the tasks we shouldn't need this. And I guess that it's better to inform `uriBaseId` than don't do it (no hard opinion).

**Comment by 3flex** [comment]
I'd like to understand more about your thinking about how path handling will be handled with the split task setup, and the design in general. I'm also not clear on the benefits of splitting it (other than for report aggregation which isn't relevant for this issue). Probably can discuss that on a separate issue. FWIW I firmly believe cacheability of all task output should be a specific goal.

To close this one in the meantime I think we can simply remove `basePath` from the set of inputs in the Gradle plugin and set basePath to the Gradle project path when running from the CLI.

**Comment by BraisGabin** [comment]
My ideas behind splitting tasks: #7674

Only talking about cacheability. If we fix #7665 then we only places where we use the `basePath` is to read the files to generate the code snippets on the md and html reports. So I guess that we could set the `basePath` as an "internal input" (I don't know the gradle terminology here) on the detekt task so it is not consider for the cache.

But we will not be able to show an absolute path on our reports any longer. For example, links.

**Comment by 3flex** [comment]
Thanks for starting the discussion - I'll chime in when time permits.

> If we fix #7665 then we only places where we use the basePath is to read the files to generate the code snippets on the md and html reports. So I guess that we could set the basePath as an "internal input"

For 2.0.0 this will work well and we can close this issue off.

**Comment by ZacSweers** [comment]
(note: some of the below details were assembled with the help of an LLM)

We reproduced this issue with Detekt 2.0.0-alpha.6 and Gradle 9.7.1. Configuration cache and Isolated Projects were enabled. We used `detekt.use.worker.api=false` and single-use Gradle daemons.

An absolute repository-root `basePath` produces correct repository-relative report paths, but moving the checkout changes the task's cache key.

We also tested the relative-path approach from [Spring Boot's workaround](https://github.com/spring-projects/spring-boot/commit/c6d66dc6ed63a92dc30e2a45a3ac957d34524ea9). It gives matching cache keys in our configuration, but the cached reports contain the producer checkout's location.

### Reproduction

We reduced the report-path behavior to a nested `:feature:sample` module. It contains one Kotlin file with a naming violation. These are the relevant tested configuration excerpts; dependency setup is omitted:

```groovy
detekt {
    toolVersion = '2.0.0-alpha.6'
    basePath.set(rootDir)
    buildUponDefaultConfig.set(true)
}

tasks.withType(dev.detekt.gradle.Detekt).configureEach {
    ignoreFailures.set(true)
    if (providers.gradleProperty('normalizeInput').isPresent()) {
        basePath.set(
            project.layout.projectDirectory.asFile.toPath().relativize(
                project.isolated.rootProject.projectDirectory.asFile.toPath()
            ).toString()
        )
    }
    reports {
        html.required.set(false)
        markdown.required.set(true)
        sarif.required.set(true)
        checkstyle.required.set(true)
    }
}
```

In this nested module, the relative value is `../..`.

1. Make two identical source checkouts at different absolute paths. Configure both to use one shared local build-cache directory.
2. Set `detekt.use.worker.api=false` in `gradle.properties`.
3. Run `:feature:sample:detekt` in the first checkout with `--no-daemon --build-cache --configuration-cache -Dorg.gradle.unsafe.isolated-projects=true -Dorg.gradle.caching.debug=true -PnormalizeInput`.
4. Run the same command in the second checkout, starting without task outputs.
5. Save the reports restored in step 4, then repeat the command with `--rerun-tasks` and compare the reports.

### Actual results

| Run | Task outcome | Report source path |
| --- | --- | --- |
| Absolute `basePath` control | Executed | Correct repository-relative path |
| Relative `basePath`, checkout A | Executed | Includes checkout A's location |
| Relative `basePath`, checkout B | `FROM-CACHE` | Still includes checkout A's location |
| Unchanged repeat in B | `UP-TO-DATE` | Still includes checkout A's location |
| Forced execution in B | Executed | Includes checkout B's location |

The relative-path runs all used cache key `c78c33265282488dde9f68f94e5571c9`. The cached and freshly generated Checkstyle and SARIF reports had different hashes in checkout B. Markdown also retained the producer checkout's location after cache restore. Launching the forced build from the module directory did not correct the paths.

The expected source path in this fixture is `feature/sample/src/main/kotlin/Example.kt`. Instead, the restored report contains a path shaped like `../../../tmp/<fixture>/first/feature/sample/src/main/kotlin/Example.kt`.

### Expected behavior

Identical sources should reuse Detekt results across checkout locations while keeping report paths relative to the intended repository root. Reports restored from cache should identify the same source files as a fresh execution in the receiving checkout.

### Implementation details

In alpha.6, [`basePath` is an optional scalar `@Input`](https://github.com/detekt/detekt/blob/v2.0.0-alpha.6/detekt-gradle-plugin/src/main/kotlin/dev/detekt/gradle/Detekt.kt#L144-L149). The plugin [passes it to the CLI unchanged](https://github.com/detekt/detekt/blob/v2.0.0-alpha.6/detekt-gradle-plugin/src/main/kotlin/dev/detekt/gradle/invoke/CliArgument.kt#L126-L128). The CLI then [converts it to an absolute path](https://github.com/detekt/detekt/blob/v2.0.0-alpha.6/detekt-cli/src/main/kotlin/dev/detekt/cli/Spec.kt#L25-L34). This suggests that the process working directory controls how the relative value is resolved. The workaround assumes resolution from the module directory.

Could the plugin track the base directory's relative location for caching and resolve the intended directory for execution? A regression test could also compare cached reports after relocation with freshly generated reports. Matching cache keys alone would miss the incorrect paths shown here.

The workaround results above cover the in-process execution mode. We haven't checked the behavior with `detekt.use.worker.api=true`.
